compiler/src/dotty/tools/dotc/transform/SyntheticMembers.scala

```diff
@@ -232,35 +232,65 @@ class SyntheticMembers(thisPhase: DenotTransformer) {
     /** The class
      *
      *  ```
-     *  package p
-     *  case class C(x: T, y: T)
+     *  case object C
+     *  ```
+     *
+     *  gets the `hashCode` method:
+     *
+     *  ```
+     *  "C".hashCode // constant folded
+     *  ```
+     *
+     *  The class
+     *
+     *  ```
+     *  case class C(x: T, y: U)
+     *  ```
+     *
+     *  if non of `T` or `U` are primitive types, gets the `hashCode` method:
+     *
+     *  ```
+     *  def hashCode: Int = ScalaRunTime._hashCode(this)
+     *  ```
+     *
+     *  else if either `T` or `U` are primitive, gets the `hashCode` method implemented by [[caseHashCodeBody]]
+     */
+    def chooseHashcode(implicit ctx: Context) = {
+      if clazz.is(ModuleClass) then
+        Literal(Constant(clazz.name.stripModuleClassSuffix.toString.hashCode))
+      else if accessors `exists` (_.info.finalResultType.classSymbol.isPrimitiveValueClass) then
+        caseHashCodeBody
+      else
+        ref(defn.ScalaRuntimeModule).select("_hashCode".toTermName).appliedTo(This(clazz))
+    }
+
+    /** The class
+     *
+     *  ```
+     *  case class C(x: Int, y: T)
      *  ```
      *
      *  gets the `hashCode` method:
      *
      *  ```
      *  def hashCode: Int = {
-     *    <synthetic> var acc: Int = "p.C".hashCode // constant folded
+     *    <synthetic> var acc: Int = 0xcafebabe
+     *    acc = Statics.mix(acc, this.productPrefix.hashCode);
      *    acc = Statics.mix(acc, x);
      *    acc = Statics.mix(acc, Statics.this.anyHash(y));
      *    Statics.finalizeHash(acc, 2)
      *  }
      *  ```
      */
     def caseHashCodeBody(implicit ctx: Context): Tree = {
-      val seed = clazz.fullName.toString.hashCode
-      if (accessors.nonEmpty) {
-        val acc = ctx.newSymbol(ctx.owner, "acc".toTermName, Mutable | Synthetic, defn.IntType, coord = ctx.owner.span)
-        val accDef = ValDef(acc, Literal(Constant(seed)))
-        val mixes = for (accessor <- accessors) yield
-          Assign(ref(acc), ref(defn.staticsMethod("mix")).appliedTo(ref(acc), hashImpl(accessor)))
-        val finish = ref(defn.staticsMethod("finalizeHash")).appliedTo(ref(acc), Literal(Constant(accessors.size)))
-        Block(accDef :: mixes, finish)
-      } else {
-        // Pre-compute the hash code
-        val hash = scala.runtime.Statics.finalizeHash(seed, 0)
-        Literal(Constant(hash))
-      }
+      val acc       = ctx.newSymbol(ctx.owner, "acc".toTermName, Mutable | Synthetic, defn.IntType, coord = ctx.owner.span)
```

Comment:
No need to align the = signs in these statements.

Reply:
The name `acc` should be define once in https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/core/StdNames.scala

Reply:
fixed in https://github.com/lampepfl/dotty/pull/7019/commits/e9ce0332806b0189aa835c55e127a6f1f0795d9d